Home Work with CompletableFuture

bannovdaniil opened the pull request:
Интересная задача, понравилась, пришлось подумать над некоторыми моментами.

bannovdaniil committed:
!1 get all directory paths

bannovdaniil committed:
!2 get all files from directory

bannovdaniil committed:
!3 частота слов в файле.

bannovdaniil committed:
!4 утилита для обработки частотности слов

bannovdaniil committed:
!4 логируемое исключение IOException

bannovdaniil committed:
!5 GoogleSearch class

bannovdaniil committed:
!6 ошибка расчета потоков поиска.

bannovdaniil committed:
!7 небольшой рефакторинг + README.md

bannovdaniil committed:
!8 рефакторинг, изменил алгоритм обработки потоков при обращении к google

bannovdaniil committed:
!9 рефакторинг кода, убрал вывод папок в которых нет файлов.

kincajou commented on parallelism/src/main/java/ru/hh/school/homework/utils/GoogleWordSearch.java:
```diff
@@ -0,0 +1,46 @@
+  private static Map<String, Long> cacheQuery = new ConcurrentHashMap<>();
+
+  public long naiveSearch(String query) {
+    if (cacheQuery.containsKey(query)) {
```
кеш тут то что нужно

kincajou commented on parallelism/src/main/java/ru/hh/school/homework/utils/GoogleWordSearch.java:
```diff
@@ -0,0 +1,46 @@
+        cacheQuery.put(query, queryFrequencies);
+      }
+      return queryFrequencies;
+    } catch (Exception err) {
```
лови только то что кидается
если там interruptedexception - ты его тут глотаешь по сути

bannovdaniil replied:
Хотел упростить себе жизнь :) их там много :) исправления внес.

kincajou commented on parallelism/src/main/java/ru/hh/school/homework/utils/GoogleWordSearch.java:
```diff
@@ -0,0 +1,46 @@
+      Document document = Jsoup.connect(GOOGLE_URL + query).userAgent(USER_AGENT).get();
+
+      long timeout = ThreadLocalRandom.current().nextLong(Constants.MIN_STEEL_TIME, Constants.MAX_STEEL_TIME);
+      Constants.LOGGER.info("Wait, google... Sleep: {}", timeout);
```
обычно не делают общую константу логгера, лучше в каждом классе свой логгер, так проще потом сконфигурить если нужно вдруг по разным файлам раскидать или другие правила логирования настроить конкретно по этому классу

kincajou commented on parallelism/src/main/java/ru/hh/school/homework/utils/DirectoriesFromPath.java:
```diff
@@ -0,0 +1,34 @@
+    this.searchPath = Paths.get(searchPath);
+  }
+
+  public void search() {
```
кажется лучше чтобы этот метод возвращал List<Path>, а getPath - ненужен

kincajou commented on parallelism/src/main/java/ru/hh/school/homework/Launcher.java:
```diff
@@ -1,83 +1,87 @@
+        CompletableFuture.runAsync(() -> frequenciesUtils.mergeWordsFrequencies(file))
+    ).toList();
+
+    CompletableFuture.allOf(cfs.toArray(new CompletableFuture[0])).join();
```
вот эта блокировка выглядит странно
как правило если ты внутри фьючи на что-то блочишься - что-то тут не так
лучше вернуть результаты allOf и сделать compose

bannovdaniil replied:
Понял учту на будущее, переделал на параллел стрим.

kincajou commented:
общий подход - если смешиваешь тред пулы (parallel стрим и свой тп) то не блочиться одним в вызове другого

bannovdaniil committed:
!10 работа над ошибками.
1.1 Исправлен Logger теперь он локальный для каждого класса где используется.
1.2 Exception в GoogleWordSearch престал быть общим
1.3 Замена блокирующего блока на parallel stream.
1.4 Рефакторинг класса DirectoriesFromPath

kincajou commented on parallelism/src/main/java/ru/hh/school/homework/Launcher.java:
```diff
@@ -1,83 +1,90 @@
-        .sorted(comparingByValue(reverseOrder()))
-        .limit(10)
-        .collect(toMap(Map.Entry::getKey, Map.Entry::getValue));
+      CompletableFuture.allOf(cfs.toArray(new CompletableFuture[0])).join();
```
стоит так же собирать и allOf-ить фьючи которые поиском занимаются
потому что конкретно вот эти - завершатся когда отработают все FilesFroDirectory.search
а поиск будет ехать дальше асинхронно

bannovdaniil replied:
Это попытался сделать.

kincajou commented on parallelism/src/main/java/ru/hh/school/homework/utils/FrequenciesUtils.java:
```diff
@@ -0,0 +1,58 @@
+          .collect(groupingBy(identity(), counting()))
+          .entrySet()
+          .stream()
+          .parallel()
```
не уверен что тут паралел чем-то поможет
померять бы в отдельном тесте

kincajou commented on parallelism/src/main/java/ru/hh/school/homework/utils/FilesFromDirectory.java:
```diff
@@ -0,0 +1,24 @@
+    List<Path> files;
+    try (var fileStream = Files.list(searchPath)) {
+      files = fileStream
+          .parallel()
```
тут тоже стоит почитать жавадок и погуглить - имеет ли смысл паралелить
возможно что работа с файловой системой не умеет биться на мелкие задачки

bannovdaniil replied:
протестировать смогу только на выходных.

kincajou replied:
в целом т.к. этот метод запускается в отдельном треде в supplyAsync то тут смысла паралелить нету - либо там последовательно, либо тут
щас получается что отдельный тред выделился и ждет пока тут паралельно (опять же - насколько паралельна работа с файловой системой?) стрим доработает

kincajou commented on parallelism/src/main/java/ru/hh/school/homework/Launcher.java:
```diff
@@ -1,83 +1,90 @@
-      .connect("https://www.google.com/search?q=" + query) //
-      .userAgent("Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/80.0.3987.116 Safari/537.36") //
-      .get();
+  private static Map<String, Long> getWordsFrequenciesList(List<Path> files) {
```
это мелкое замечание, по общему стилю
не очень зашло какие методы где лежат и к чему относятся, какой код про какой знает

тут либо подход - несколько методов в общем классе Launcher, без объектной структуры, что вполне ок для данной задачи
либо, если хочется все по классам раскидать, то я бы ждал что

new DirectoriesFromPath(Constants.ROOT_DIRECTORY).searchAndLog() все делает внутри
и вызывает уже FilesFromDirectory, а тот уже умеет из файлов почитать и в поиск сходить

bannovdaniil replied:
Это я про SOLID начитался, пытался реализовать.

bannovdaniil committed:
!11 работа над ошибками: добавил AllOff на внутренний CompletableFuture

bannovdaniil committed:
!12 работа над ошибками: изменил AllOff на thenCompose ...

kincajou commented on parallelism/src/main/java/ru/hh/school/homework/Launcher.java:
```diff
@@ -1,83 +1,89 @@
+    try (ExecutorService executorService = Executors.newFixedThreadPool(Constants.MAX_SEARCH_THREAD)) {
+      List<CompletableFuture<Void>> cfs =
+          paths.stream().map(path ->
+              CompletableFuture.supplyAsync(() -> FilesFromDirectory.search(path, Constants.EXTENSION))
```
вот тут будет выполняться в системном fork join pool который по количеству цпу, лучше иметь отдельный тп побольше

kincajou commented on parallelism/src/main/java/ru/hh/school/homework/Launcher.java:
```diff
@@ -1,83 +1,89 @@
+              CompletableFuture.supplyAsync(() -> FilesFromDirectory.search(path, Constants.EXTENSION))
+                  .thenCompose(fileList -> {
+                    LOGGER.info("Path: {}, files: {}", path, fileList.size());
+                    return CompletableFuture.supplyAsync(() -> getWordsFrequenciesList(fileList))
```
тут тоже системный fjp

kincajou commented on parallelism/src/main/java/ru/hh/school/homework/Launcher.java:
```diff
@@ -1,83 +1,89 @@
+                  .thenCompose(fileList -> {
+                    LOGGER.info("Path: {}, files: {}", path, fileList.size());
+                    return CompletableFuture.supplyAsync(() -> getWordsFrequenciesList(fileList))
+                        .thenAcceptAsync(frequenciesList -> showWordsFrequencies(path, frequenciesList), executorService);
```
тут похожая ситуация - запускается тред и стоит ждет когда внутри showWordsFrequencies доработает паралел стрим

kincajou commented:
финалим

первая часть по подсчету слов хорошо бы решилась через fork join pool (без наворачивания сверху CompletableFuture) - либо через RecursiveTask с форком на папках и файлах и подсчете топа при джоине, либо на паралел стримах, но там рекурсии не получится

фтора часть с подсчетом - completable future норм

executor стоит все же гасить самому, при этом дождавшись когда доработают треды из первой и второй частей, для CF это собрать все фьючи и сделать на них allOf(...).join

стоит разобраться с наслоением CF а внутри parallel stream - такого быть не должно
и еще раз подумать про разбиение на классы

спасибо за домашку!
